rclcpp/include/rclcpp/timer.hpp

```diff
@@ -101,16 +108,19 @@ class TimerBase
    * The multithreaded executor takes advantage of this to avoid scheduling
    * the callback multiple times.
    *
-   * \return `true` if the callback should be executed, `false` if the timer was canceled.
+   * \return a shared_ptr if the callback should be executed, std::nullopt if the timer was canceled.
    */
   RCLCPP_PUBLIC
-  virtual bool
+  virtual std::optional<std::shared_ptr<void>>
```

Comment:
I'm a little confused as to why you did `std::optional<std::shared_ptr<void>>` here.  Wouldn't `std::shared_ptr<void>` be sufficient?

Reply:
Yes a std::shared_ptr containing a nullptr would work, but in my opinion the semantics are nicer if using an optional.

Reply:
I think that having a `std::shared_ptr<void>` is better for consistency.

Other pieces of code added in this PR (like the `struct ExecutorEvent`) contain a `std::shared_ptr<void> data;` that could be a nullptr, thus we already need to check that pointer before using it.
Having this API using also an optional, seems an unnecessary additional layer

Reply:
changes it to std::shared_ptr<void>